src/main/java/htsjdk/beta/io/IOPathUtils.java

```diff
@@ -69,4 +71,45 @@ public static void writeStringToPath(final IOPath ioPath, final String contents)
                     e);
         }
     }
+
+    /**
+     * Takes an IOPath and returns a new IOPath object that keeps the same basename as the original but has
+     * a new extension. If append is set to false, only the last component of an extension will be replaced.
+     * e.g. "my.fasta.gz" -> "my.fasta.tmp"
+     *
+     * If the input IOPath was created from a rawInputString that specifies a relative local path, the new path will
+     * have a rawInputString that specifies an absolute path.
+     *
+     * Examples:
+     *     - (test_na12878.bam, .bai) -> test_na12878.bai (append = false)
+     *     - (test_na12878.bam, .md5) -> test_na12878.bam.md5 (append = true)
+     *
+     * @param path The original path
+     * @param newExtension A new file extension. Must include the leading dot e.g. ".txt", ".bam"
+     * @param append If set to true, append the new extension to the original basename. If false, replace the original extension
+     *               with the new extension. If append = false and the original name has no extension, an exception will be thrown.
+     * @param ioPathConstructor a function that takes a string and returns an IOPath-derived class of type <T>
+     * @return A new IOPath object with the new extension
+     */
+    public static <T extends IOPath> T replaceExtension(
```

Comment:
Yeah, I like that better. I back ported this from one of Takuto's Picard PRs since I needed it. Also, I made him add tests for it ;-). Will separate these and add some tests.

Reply:
Oh, wait I *did* add tests. Good for me. Now I have to update them.

Reply:
Done.